src/libopenrave/kinbodyjoint.cpp

```diff
@@ -1817,83 +1845,156 @@ void KinBody::Joint::SetMimicEquations(int iaxis, const std::string& poseq, cons
+            const bool bfound = keyvalue.first.first == thisdofformat;
+#if defined(DEBUG_KINBODYJOINT_CALCULATEJACOBIAN)
+            if(bfound) {
+                RAVELOG_VERBOSE_FORMAT("Found cached %s", thisdofformat.to_string());
             }
-            dReal fvel = _vmimic[iaxis]->_velfns.at(itmimicdof->dofformatindex)->Eval(vtempvalues.empty() ? NULL : &vtempvalues[0]);
-            const MIMIC::DOFFormat& dofformat = _vmimic[iaxis]->_vdofformat.at(itmimicdof->dofformatindex);
-            if( dofformat.GetJoint(*parent)->IsMimic(dofformat.axis) ) {
-                dofformat.GetJoint(*parent)->_ComputePartialVelocities(vtemppartials,dofformat.axis,mapcachedpartials);
-                dReal fpartial = 0;
-                FOREACHC(itpartial,vtemppartials) {
-                    if( itpartial->first == itmimicdof->dofindex ) {
-                        fpartial += itpartial->second;
-                    }
-                }
-                fvel *= fpartial;
-            }
-            // before pushing back, check for repetition
-            bool badd = true;
-            FOREACH(itpartial,vpartials) {
-                if( itpartial->first == itmimicdof->dofindex ) {
-                    itpartial->second += fvel;
-                    badd = false;
-                    break;
-                }
+#endif
+            return bfound;
+        }
+    );
+    const bool bCached = mit != end(mTotalderivativepairValue);
+    if(bCached) {
+        for(const std::pair<const std::pair<Mimic::DOFFormat, int>, dReal> &keyvalue : mTotalderivativepairValue) {
+            if(keyvalue.first.first == thisdofformat) {
+                const int dependedJointIndex = keyvalue.first.second;
+                const dReal partialDerivative = keyvalue.second;
+                vDofindexDerivativePairs.emplace_back(dependedJointIndex, partialDerivative); // collect all dz/dx
             }
-            if( badd ) {
-                vpartials.emplace_back(itmimicdof->dofindex,  fvel);
+        }
+#if defined(DEBUG_KINBODYJOINT_CALCULATEJACOBIAN)
+        RAVELOG_VERBOSE_FORMAT("mTotalderivativepairValue: %s\nvDofindexDerivativePairs: %s", to_string(mTotalderivativepairValue) % to_string(vDofindexDerivativePairs));
+#endif
+        return;
+    }
+
+    /* ========== Collect values of joints on which joint z depends ========== */
+    const MimicPtr& pmimic = _vmimic[iaxis];
+    const std::vector<Mimic::DOFFormat>& vdofformats = pmimic->_vdofformat; ///< collection of information of joints
+
+    std::vector<dReal> vDependedJointValues; ///< collect values of joints on which this joint *directly* depends on
+    for(const Mimic::DOFFormat& dofformat : vdofformats) {
+        const JointConstPtr dependedjoint = dofformat.GetJoint(*parent); ///< say joint y
+        vDependedJointValues.push_back(dependedjoint->GetValue(dofformat.axis));
+    }
+
+#if defined(DEBUG_KINBODYJOINT_CALCULATEJACOBIAN)
```

Comment:
@tgn3000 can't you just do `if( IS_DEBUGLEVEL(Level_Verbose) )`?

Reply:
ok